rmw_connext_cpp/src/functions.cpp

```diff
@@ -164,62 +212,73 @@ rmw_create_publisher(
     rmw_set_error_string("callbacks handle is null");
     return NULL;
   }
-  std::string type_name = std::string(callbacks->package_name) + "::msg::dds_::" +
-    callbacks->message_name + "_";
+  std::string type_name = _create_type_name(callbacks, "msg");
+  // Past this point, a failure results in unrolling code in the goto fail block.
```

Comment:
Based on that link I'd say `goto fail` is more obvious, doesn't have problems when inside nesting loops, and can be expanded to have multiple recovery stages with different labels. Where as the `do {} while(false)` method can have problems when the interior code needs to break out of a loop but not for an error and it cannot have multiple recovery labels but instead requires conditionals to be set in the interior and evaluated in the recovery code.

So I'd say there's not sufficient reason to change away from the `goto fail`. Obligatory:

https://xkcd.com/292/